File: src/main/java/com/polygloat/controllers/LanguageController.java
package com.polygloat.controllers;

import com.polygloat.constants.Message;
import com.polygloat.dtos.request.LanguageDTO;
import com.polygloat.dtos.request.validators.LanguageValidator;
import com.polygloat.exceptions.NotFoundException;
import com.polygloat.model.Language;
import com.polygloat.model.Permission;
import com.polygloat.model.Repository;
import com.polygloat.service.LanguageService;
import com.polygloat.service.RepositoryService;
import com.polygloat.service.SecurityService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.LinkedHashSet;
import java.util.Set;
import java.util.stream.Collectors;

@RestController
@CrossOrigin(origins = "*")
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@RequestMapping("/api/repository/{repositoryId}/languages")
public class LanguageController implements IController {
    private final LanguageService languageService;
    private final RepositoryService repositoryService;
    private final LanguageValidator languageValidator;
    private final SecurityService securityService;

    @PostMapping(value = "")
    public LanguageDTO createLanguage(@PathVariable("repositoryId") Long repositoryId,
                                      @RequestBody @Valid LanguageDTO dto) {
        Repository repository = repositoryService.findById(repositoryId).orElseThrow(NotFoundException::new);
        securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.MANAGE);
        languageValidator.validateCreate(dto, repository);
        Language language = languageService.createLanguage(dto, repository);
        return LanguageDTO.fromEntity(language);
    }

    @PostMapping(value = "/edit")
    public LanguageDTO editLanguage(@RequestBody @Valid LanguageDTO dto) {
        languageValidator.validateEdit(dto);
        Language language = languageService.findById(dto.getId()).orElseThrow(() -> new NotFoundException(Message.LANGUAGE_NOT_FOUND));
        securityService.checkRepositoryPermission(language.getRepository().getId(), Permission.RepositoryPermissionType.MANAGE);
        return LanguageDTO.fromEntity(languageService.editLanguage(dto));
    }

    @GetMapping(value = "")
    public Set<LanguageDTO> getAll(@PathVariable("repositoryId") Long repositoryId) {
        return languageService.findAll(repositoryId).stream().map(LanguageDTO::fromEntity)
                .collect(Collectors.toCollection(LinkedHashSet::new));
    }

    @GetMapping(value = "{id}")
    public LanguageDTO get(@PathVariable("id") Long id) {
        Language language = languageService.findById(id).orElseThrow(NotFoundException::new);
        securityService.getAnyRepositoryPermission(language.getRepository().getId());
        return LanguageDTO.fromEntity(language);
    }

    @DeleteMapping(value = "/{id}")
    public void deleteLanguage(@PathVariable Long id) {
        Language language = languageService.findById(id).orElseThrow(() -> new NotFoundException(Message.LANGUAGE_NOT_FOUND));
        securityService.checkRepositoryPermission(language.getRepository().getId(), Permission.RepositoryPermissionType.MANAGE);
        languageService.deleteLanguage(id);
    }
}

File: src/main/kotlin/io/tolgee/controllers/LanguageController.kt
@file:Suppress("MVCPathVariableInspection")

package io.tolgee.controllers

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import io.swagger.v3.oas.annotations.tags.Tags
import io.tolgee.constants.Message
import io.tolgee.dtos.request.LanguageDto
import io.tolgee.dtos.request.validators.LanguageValidator
import io.tolgee.exceptions.NotFoundException
import io.tolgee.model.Permission
import io.tolgee.security.AuthenticationFacade
import io.tolgee.security.api_key_auth.AccessWithApiKey
import io.tolgee.service.LanguageService
import io.tolgee.service.ProjectService
import io.tolgee.service.SecurityService
import org.springframework.web.bind.annotation.*
import java.util.stream.Collectors
import javax.validation.Valid

@RestController
@CrossOrigin(origins = ["*"])
@RequestMapping(value = [
    "/api/project/{projectId:[0-9]+}/languages",
    "/api/project/languages"
])
@Tags(value = [
    Tag(name = "Languages", description = "Languages"),
])
open class LanguageController(
        private val languageService: LanguageService,
        private val projectService: ProjectService,
        private val languageValidator: LanguageValidator,
        private val securityService: SecurityService,
        private val authenticationFacade: AuthenticationFacade
) : IController {

    @PostMapping(value = [""])
    @Operation(summary = "Creates language")
    fun createLanguage(@PathVariable("projectId") projectId: Long,
                       @RequestBody @Valid dto: LanguageDto?): LanguageDto {
        val project = projectService.get(projectId).orElseThrow { NotFoundException() }
        securityService.checkProjectPermission(projectId, Permission.ProjectPermissionType.MANAGE)
        languageValidator.validateCreate(dto, project)
        val language = languageService.createLanguage(dto, project!!)
        return LanguageDto.fromEntity(language)
    }

    @Operation(summary = "Edits language")
    @PostMapping(value = ["/edit"])
    fun editLanguage(@RequestBody @Valid dto: LanguageDto?): LanguageDto {
        languageValidator.validateEdit(dto)
        val language = languageService.findById(dto!!.id!!).orElseThrow { NotFoundException(Message.LANGUAGE_NOT_FOUND) }
        securityService.checkProjectPermission(language.project!!.id, Permission.ProjectPermissionType.MANAGE)
        return LanguageDto.fromEntity(languageService.editLanguage(dto))
    }

    @GetMapping(value = [""])
    @AccessWithApiKey
    @Operation(summary = "Returns all project languages", tags = ["API KEY", "Languages"])
    fun getAll(@PathVariable("projectId") pathProjectId: Long?): Set<LanguageDto> {
        val projectId = if (pathProjectId === null) authenticationFacade.apiKey.project!!.id else pathProjectId
        securityService.checkAnyProjectPermission(projectId)
        return languageService.findAll(projectId).stream().map { LanguageDto.fromEntity(it) }
                .collect(Collectors.toCollection { LinkedHashSet() })
    }

    @GetMapping(value = ["{id}"])
    @Operation(summary = "Returns specific language")
    operator fun get(@PathVariable("id") id: Long?): LanguageDto {
        val language = languageService.findById(id!!).orElseThrow { NotFoundException() }
        securityService.checkAnyProjectPermission(language.project!!.id)
        return LanguageDto.fromEntity(language)
    }

    @Operation(summary = "Deletes specific language")
    @DeleteMapping(value = ["/{id}"])
    fun deleteLanguage(@PathVariable id: Long) {
        val language = languageService.findById(id).orElseThrow { NotFoundException(Message.LANGUAGE_NOT_FOUND) }
        securityService.checkProjectPermission(language.project!!.id, Permission.ProjectPermissionType.MANAGE)
        languageService.deleteLanguage(id)
    }
}
